main/src/library/Fixpoint/Solver.flix

```diff
@@ -399,44 +399,44 @@ namespace Fixpoint {
     /// Returns all facts in `d` associated with the predicate symbol `p`.
     ///
     pub def facts1(p: PredSym, d: Datalog[Boxed]): Array[v] with Boxable[v] =
-        let f = terms -> unbox(terms[0]);
+        let f = terms -> unbox(headTermValue(terms[0]));
         factsOf(f, p, d) as & Pure
 
     ///
     /// Returns all facts in `d` associated with the predicate symbol `p`.
     ///
     pub def facts2(p: PredSym, d: Datalog[Boxed]): Array[(t1, t2)] with Boxable[t1], Boxable[t2] =
-        let f = terms -> (unbox(terms[0]), unbox(terms[1]));
+        let f = terms -> (unbox(headTermValue(terms[0])), unbox(headTermValue(terms[1])));
         factsOf(f, p, d) as & Pure
 
     ///
     /// Returns all facts in `d` associated with the predicate symbol `p`.
     ///
     pub def facts3(p: PredSym, d: Datalog[Boxed]): Array[(t1, t2, t3)] with Boxable[t1], Boxable[t2], Boxable[t3] =
-        let f = terms -> (unbox(terms[0]), unbox(terms[1]), unbox(terms[2]));
+        let f = terms -> (unbox(headTermValue(terms[0])), unbox(headTermValue(terms[1])), unbox(headTermValue(terms[2])));
         factsOf(f, p, d) as & Pure
 
     ///
     /// Returns all facts in `d` associated with the predicate symbol `p`.
     ///
     pub def facts4(p: PredSym, d: Datalog[Boxed]): Array[(t1, t2, t3, t4)] with Boxable[t1], Boxable[t2], Boxable[t3], Boxable[t4] =
-        let f = terms -> (unbox(terms[0]), unbox(terms[1]), unbox(terms[2]), unbox(terms[3]));
+        let f = terms -> (unbox(headTermValue(terms[0])), unbox(headTermValue(terms[1])), unbox(headTermValue(terms[2])), unbox(headTermValue(terms[3])));
         factsOf(f, p, d) as & Pure
 
     ///
     /// Returns all facts in `d` associated with the predicate symbol `p`.
     ///
     pub def facts5(p: PredSym, d: Datalog[Boxed]): Array[(t1, t2, t3, t4, t5)] with Boxable[t1], Boxable[t2], Boxable[t3], Boxable[t4], Boxable[t5] =
-        let f = terms -> (unbox(terms[0]), unbox(terms[1]), unbox(terms[2]), unbox(terms[3]), unbox(terms[4]));
+        let f = terms -> (unbox(headTermValue(terms[0])), unbox(headTermValue(terms[1])), unbox(headTermValue(terms[2])), unbox(headTermValue(terms[3])), unbox(headTermValue(terms[4])));
         factsOf(f, p, d) as & Pure
 
     ///
     /// Returns an array of facts associated with the given predicate symbol `p` in the given Datalog program `d`.
     ///
-    def factsOf(f: Array[v] -> t & e, p: PredSym, d: Datalog[v]): Array[t] & Impure =
+    def factsOf(f: Array[HeadTerm[v]] -> t & e, p: PredSym, d: Datalog[v]): Array[t] & Impure =
         let Datalog(facts, _) = d;
         restrict(facts, Set#{p}) |>
```

Comment:
Presumably restrict traverses all facts once and filters them. Perhaps `filterMap` can be used. Or alternatively, for even more speed, you could use `Array.foreach` and a `MutList` to accumulate the result?

Reply:
Yes, both are possible. However `Array.filterMap` has not been optimized like `Array.filter`. Why do you expect `MutList` to be faster than a (optimized) `filterMap`?

Reply:
Only because it avoids lots of allocation of `Some` objects.

Reply:
Fair point. Would `MutList.reserve!` make sense to have an array that's large enough upfront or would that be too space inefficient? It would have same length as `facts`.